File: convert-sorted-array-to-bst.cc
#include <cstdlib>
#include <vector>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *sortedArrayToBST(vector<int> &num) {
    return array_to_bst(num, 0, num.size() - 1);
  }

  TreeNode* array_to_bst(vector<int>& num, int l, int r) {
    if (l > r) return NULL;
    int mid = (l + r) / 2;
    TreeNode* root = new TreeNode(num[mid]);
    root->left = array_to_bst(num, l, mid - 1);
    root->right = array_to_bst(num, mid + 1, r);
    return root;
  }
};

File: convert-sorted-array-to-bst.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedArrayToBST(int[] num) {
        return convert(num, 0, num.length - 1);
    }

    private TreeNode convert(int[] num, int l, int r) {
        if (l > r) return null;
        if (l == r) return new TreeNode(num[l]);
        int mid = (l + r) / 2;
        TreeNode node = new TreeNode(num[mid]);
        node.left = convert(num, l, mid - 1);
        node.right = convert(num, mid + 1, r);
        return node;
    }
}
